### PR TITLE
feat: New Handshake based on Noise IK pattern

### PR DESCRIPTION
**Pull request by goldroom**
New Handshake based on Noise IK pattern

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1561)

### CODE REVIEW

**Comment by gjedeer** [comment]
This will be a backwards-incompatible change, yes? I have a backwards-incompatible change coming up in tuntox and would like to schedule it at the same time to minimize the chaos.

**Comment by zoff99** [comment]
@gjedeer yes how it looks now it will not be compatible with current toxcore. also NGC (new group chats) will be a breaking change.
we do not yet have an exact date when these 2 will be merged into master.

File: CMakeLists.txt [hunk 1]
```diff
@@ -170,7 +170,8 @@ set(toxcore_SOURCES ${toxcore_SOURCES}
 include(CheckFunctionExists)
 check_function_exists(explicit_bzero HAVE_EXPLICIT_BZERO)
 check_function_exists(memset_s HAVE_MEMSET_S)
-set(toxcore_LINK_MODULES ${toxcore_LINK_MODULES} ${LIBSODIUM_LIBRARIES})
+set(toxcore_LINK_MODULES ${toxcore_LINK_MODULES} ${LIBSODIUM_LIBRARIES} "/usr/local/lib/libnoiseprotocol.a")
```
Comment (iphydf) [inline CMakeLists.txt:173]: We'll need to search for the library instead of hard-coding it here.

File: toxcore/net_crypto.c [hunk 2]
```diff
@@ -40,23 +43,32 @@ typedef enum Crypto_Conn_State {
...
 } Crypto_Conn_State;
 
 typedef struct Crypto_Connection {
     uint8_t public_key[CRYPTO_PUBLIC_KEY_SIZE]; /* The real public key of the peer. */
-    uint8_t recv_nonce[CRYPTO_NONCE_SIZE]; /* Nonce of received packets. */
-    uint8_t sent_nonce[CRYPTO_NONCE_SIZE]; /* Nonce of sent packets. */
-    uint8_t sessionpublic_key[CRYPTO_PUBLIC_KEY_SIZE]; /* Our public key for this session. */
-    uint8_t sessionsecret_key[CRYPTO_SECRET_KEY_SIZE]; /* Our private key for this session. */
-    uint8_t peersessionpublic_key[CRYPTO_PUBLIC_KEY_SIZE]; /* The public key of the peer. */
+    //AKE NEW: nonces not needed for Noise => handshake object
```
Comment (iphydf) [inline toxcore/net_crypto.c:53]: These commented out things need to be cleaned up.

How much work would it be to make the new protocol optional (either by compile time flag or a Tox_Options field)?

File: toxcore/net_crypto.c [hunk 3]
```diff
@@ -446,48 +525,206 @@ static int handle_cookie_response(const Logger *log, uint8_t *cookie, uint64_t *
...
+        //AKE NEW: Points to the message buffer to be populated with handshake details and the message payload.
+        noise_buffer_set_output(noise_message, noise_message_buf, sizeof(noise_message_buf));
+        //AKE NEW: Computes Noise message B pattern "e, ee, se"
+        int err = noise_handshakestate_write_message(handshake, &noise_message, &noise_payload);
 
-    random_nonce(packet + 1 + COOKIE_LENGTH);
-    int len = encrypt_data(peer_real_pk, c->self_secret_key, packet + 1 + COOKIE_LENGTH, plain, sizeof(plain),
-                           packet + 1 + COOKIE_LENGTH + CRYPTO_NONCE_SIZE);
+        if (err != NOISE_ERROR_NONE) {
+            noise_perror("write handshake", err);
```
Comment (iphydf) [inline toxcore/net_crypto.c:611]: This probably needs to become a LOGGER statement so we don't write to stderr from a library.

File: toxcore/net_crypto.c [hunk 4]
```diff
@@ -505,18 +742,41 @@ static int create_crypto_handshake(const Net_Crypto *c, uint8_t *packet, const u
...
+                                   uint8_t *dht_public_key, uint8_t *cookie, const uint8_t *packet, uint16_t length, const uint8_t *expected_real_pk,
+                                   NoiseHandshakeState *handshake)
 {
-    if (length != HANDSHAKE_PACKET_LENGTH) {
+	//AKE NEW: Debug output
+//    fprintf(stderr, "ENTERING: handle_crypto_handshake()\n");
+    //AKE NEW: get handhake role => different handshake packet lengths
+    int role = noise_handshakestate_get_role(handshake);
+
+    if (role == NOISE_ROLE_RESPONDER) {
```
Comment (iphydf) [inline toxcore/net_crypto.c:761]: This whole nested if/else statement can be a single if.

File: toxcore/net_crypto.c [hunk 5]
```diff
@@ -658,6 +1008,80 @@ static IP_Port return_ip_port_connection(Net_Crypto *c, int crypt_connection_id)
...
+    } else {
+        fprintf(stderr, "Client private key required, but not provided.\n");
+        return -1;
+    }
+
+    //AKE NEW: IK pattern: if role = INITIATOR => peer_public_key is also necessary
+    int role = noise_handshakestate_get_role(handshake);
+
+    if (role == NOISE_ROLE_INITIATOR) {
+        if (peer_public_key) {
```
Comment (iphydf) [inline toxcore/net_crypto.c:1062]: Invert check to return early and reduce nesting.

File: toxcore/net_crypto.c [hunk 6]
```diff
@@ -1943,14 +2625,60 @@ static int handle_new_connection_handshake(Net_Crypto *c, IP_Port source, const
...
+    //AKE NEW: initialize_handshake
+    //AKE NEW: IMPORTANT n_c.publicy_key is empty/not defined here, it's set handle_crypto_handshake()...
+    if (initialize_handshake(n_c.handshake_temp, c->self_secret_key, n_c.public_key) == -1) {
+        noise_handshakestate_free(n_c.handshake_temp);
+        return -1;
+    }
+
+    int role = noise_handshakestate_get_role(n_c.handshake_temp);
+
+    if (role == NOISE_ROLE_RESPONDER) {
```
Comment (iphydf) [inline toxcore/net_crypto.c:2655]: Invert condition/return early.

**Review by iphydf — CHANGES_REQUESTED** [review]
This code needs a bit of cleanup and testing. Can you fix the CI builds? We'll need to install the noise library there and find it with cmake. I'll take care of the bazel build once that's done.

**Comment by goldroom** [comment]
Closing in favor of https://github.com/TokTok/c-toxcore/pull/2450.

**Pull request closed without merging**